### PR TITLE
Upgrade webpack v5 and related modules

### PR DESCRIPTION
**Pull request by thedavidprice**
I know, I know... Webpack v5 takes longer to build. I'd argue it's time to upgrade as other packages are deprecating v4. And this will give us impetus to push forward with ESbuild.

https://webpack.js.org/migrate/5/

## To Do:
- [x] resolve `yarn rw dev` start error
- [ ] needs testing with more complex apps
- [x] [Added to this PR] depends on upgrading to [Storybook >= v6.2](https://github.com/storybookjs/storybook/releases/tag/v6.2.0) (#2370) and opting-in: https://storybook.js.org/blog/storybook-for-webpack-5/
- [ ] Test with ESBuild ([see forum post](https://community.redwoodjs.com/t/faster-compilation-times-with-esbuild-in-v0-28-0/1979))
- [x] [See comments below] Also confirm working with [error-overlay-webpack-plugin](https://github.com/smooth-code/error-overlay-webpack-plugin)
   - related PR with support for v5: https://github.com/smooth-code/error-overlay-webpack-plugin/pull/72

Starting the Webpack Dev Server includes the following error (last line starting with `web | [HPM] Error`). However, this does not seem to actually affect the Dev Server running and the error is not repeated:
```
...
api | Running at 'http://localhost:8911'
api | Watching files in '/Users/price/Repos/tdp-redwood-tutorial-test/api/src/functions'
web | ℹ ｢wds｣: Project is running at http://localhost:8910/
web | ℹ ｢wds｣: webpack output is served from /
web | ℹ ｢wds｣: Content not from webpack is served from /Users/price/Repos/tdp-redwood-tutorial-test/web
web | ℹ ｢wds｣: 404s will fallback to /index.html
...
web | [HPM] Error occurred while trying to proxy request /graphql from localhost:8910 to http://[::1]:8911 (ECONNRESET) (https://nodejs.org/api/errors.html#errors_common_system_errors)
...
```

## Possible improvement:
-[ ] "If you have rules defined for loading assets using raw-loader, url-loader, or file-loader, please use Asset Modules instead as they're going to be deprecated in near future.": https://webpack.js.org/guides/asset-modules/

### CODE REVIEW

File: packages/core/package.json [hunk 1]
```diff
@@ -35,46 +35,46 @@
...
     "lodash.escaperegexp": "^4.1.2",
-    "mini-css-extract-plugin": "^1.4.1",
+    "mini-css-extract-plugin": "^1.5.1",
     "null-loader": "^4.0.1",
     "prisma": "2.21.2",
     "react-refresh": "^0.10.0",
     "style-loader": "^1.1.3",
     "svg-react-loader": "^0.4.6",
-    "terser-webpack-plugin": "^4.2.3",
+    "terser-webpack-plugin": "^5.1.1",
```
Comment (dac09) [inline packages/core/package.json:68]: I think webpack 5 includes terser by default. We can probably remove this and in the optimizations block (in webpack.common.js) do this:

```
    minimizer: [new CssMinimizer(), '...'],
```
The `...` is the new supported syntax

File: packages/core/package.json [hunk 2]
```diff
@@ -35,46 +35,46 @@
...
-    "webpack-merge": "^5.1.2",
-    "webpack-retry-chunk-load-plugin": "^1.4.0",
+    "webpack": "^5.36.1",
+    "webpack-bundle-analyzer": "^4.4.1",
+    "webpack-cli": "^4.6.0",
+    "webpack-dev-server": "^3.11.2",
+    "webpack-manifest-plugin": "^3.1.1",
+    "webpack-merge": "^5.7.3",
+    "webpack-retry-chunk-load-plugin": "^2.2.0",
     "whatwg-fetch": "^3.5.0"
```
Comment (dac09) [inline packages/core/package.json:78]: "@pmmmwh/react-refresh-webpack-plugin": "^0.4.3", and react-refresh probably need bumping too

File: packages/core/package.json (hunk 1 above)
Comment (thedavidprice) [reply packages/core/package.json:68]: https://www.npmjs.com/package/terser-webpack-plugin

done

File: packages/core/package.json (hunk 2 above)
Comment (thedavidprice) [reply packages/core/package.json:78]: done

Comment (dac09) [reply packages/core/package.json:78]: Think you forgot to push :P

**Comment by thedavidprice** [comment]
## Progress So Far
**tl;dr:** I think we're close!

After following numerous threads, I've managed to get the E2E Tutorial tests passing and working well! The current cost of moving to v5 is that several packages/plugins related to `react-error-overlay` are not working with v5 and the associated `webpack-dev-server` v4-beta. (Note: Webpack v5 + DevServer v3 is broken. It's recommend for all users of DevServer to use the beta. And there were a _lot_ of breaking API changes... not super fun.)
- `webpack-dev-server` [Release Notes](https://github.com/webpack/webpack-dev-server/releases)

I've also followed the [guide for Storybook + Webpack v5 support](https://gist.github.com/shilman/8856ea1786dcd247139b47b270912324#upgrade). See below for current status.

### Disabled Error Overlay
The new DevServer v4 has the built in React error overlay enabled by default. It's not spectacular, but it works (screenshot at bottom). 

I've disabled the React-refresh overlay in `webpack.common.js`, which I don't believe we needed anyway as it was redundant:
```js
      new ReactRefreshWebpackPlugin({ overlay: false }), // bug with Webpack v5
```

I also had to disable `error-overlay-webpack-plugin` in `webpack.development.js`, which is a bummer because it is an [improvement compared to the default display](https://github.com/smooth-code/error-overlay-webpack-plugin). Currently I've just uncommented the lines of code. We should discuss if we want to leave it as is or wait for v5 support or just copy+paste the plugin into our project and update the API changes (I know what they are now). The project is sitting with on a 30-day open PR with no maintainer response. Not great.

**TO DO**
- [ ] Discuss next steps with React Error Overlay Issues


## Prerender Failing E2E
@dac09 using `yarn rwt link ...` and my Tutorial Blog test project, I can _successfully_ run build with pretender!
Screen Shot 2021-05-19 at 6 39 40 PM

However, the E2E is failing on pretender although it's succeeding on build. I'm not seeing anything helpful in the stack trace and Cypress cloud locked us out for the month.  Any chance you could take a look and help me understand what's going on?

**TO DO**
- [ ] Resolve `yarn rw build --prerender` and confirm it's working locally (just needs E2E TLC)

## Storybook fails on directory error
@peterp When I run `yarn rw storybook` I make it past the build but then hit this error message:
```
...
<i> [webpack-dev-middleware] wait until bundle finished
10% building 0/1 entries 0/0 dependencies 0/0 modulesinfo => Using cached manager
99% done plugins webpack-hot-middlewarewebpack built preview 3f0676d3cfe8f943b87b in 2077ms
ModuleBuildError: Module build failed (from ../node_modules/babel-loader/lib/index.js):
Error: ENOENT: no such file or directory, open '/Users/price/Repos/tdp-redwood-tutorial-test/redwood/core/config/storybook/generated-stories-entry.js'
    at processResult (/Users/price/Repos/tdp-redwood-tutorial-test/node_modules/webpack/lib/NormalModule.js:700:19)
    at /Users/price/Repos/tdp-redwood-tutorial-test/node_modules/webpack/lib/NormalModule.js:806:5
    at /Users/price/Repos/tdp-redwood-tutorial-test/node_modules/loader-runner/lib/LoaderRunner.js:399:11
    at Array.<anonymous> (/Users/price/Repos/tdp-redwood-tutorial-test/node_modules/loader-runner/lib/LoaderRunner.js:221:19)
    at runCallbacks (/Users/price/Repos/tdp-redwood-tutorial-test/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:27:15)
    at ReadFileContext.<anonymous> (/Users/price/Repos/tdp-redwood-tutorial-test/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:200:4)
    at ReadFileContext.callback (/Users/price/Repos/tdp-redwood-tutorial-test/node_modules/graceful-fs/graceful-fs.js:123:16)
    at FSReqCallback.readFileAfterOpen [as oncomplete] (fs.js:273:13)
ModuleBuildError: Module build failed (from ../node_modules/babel-loader/lib/index.js):
Error: ENOENT: no such file or directory, open '/Users/price/Repos/tdp-redwood-tutorial-test/redwood/core/config/storybook/preview.js-generated-config-entry.js'
    at processResult (/Users/price/Repos/tdp-redwood-tutorial-test/node_modules/webpack/lib/NormalModule.js:700:19)
...
```

I've followed [this guide](https://gist.github.com/shilman/8856ea1786dcd247139b47b270912324#upgrade) — perhaps I didn't implement it correctly? And/or I might not have made the breaking change updates to the storybook-specific webpack config. Could you possibly take a look and give me some direction about where/how to hunt this down?

**TO DO**
- [ ] Resolve `yarn rw storybook` build error

## Double '"ctrl+c" required to kill DevServer process
@peterp When I try to kill the `yarn rw dev` process, I'm seeing this console output, which requires a second "ctl+c" to successfully kill the Redwood dev server:
```
cd "/Users/price/Repos/tdp-redwood-tutorial-test/web" && yarn webpack serve --config ../node_modules/@redwoodjs/core/config/webpack.development.js  exited with code SIGINT
```

Any suggestions or strategies about resolving something like this?

**TO DO**
- [ ] resolve the double "ctrl+c" to kill Redwood Dev Server

---

### Default React Error Overlay Screenshot
Screen Shot 2021-05-19 at 6 22 01 PM

**Comment by thedavidprice** [comment]
@jtoar this might end up requiring PostCSS v8. Possible to work on upgrading that in parallel?

**Comment by jtoar** [comment]
@thedavidprice Indeed, I'm long overdue. Prioritizing that now

**Comment by jtoar** [comment]
@thedavidprice spilling my process so far. After searching the package.jsons and using a combination of `yarn list` and `yarn why`, I'm not sure that we actually have postcss installed _directly_ (i.e. there's nothing to update). But one of the things that was holding us back from using postcss 8 via the postcss-loader (for the tailwind setup command) was css-loader. The version of css-loader we were using before this PR was `^4.2.2`. css-loader only upgraded to postcss 8 in v5 (see the release notes here: https://github.com/webpack-contrib/css-loader/releases/tag/v5.0.0); your PR here puts us at `^5.2.5`, the latest.

I used your PR with yarn rwt link on a new project and installed the latest versions of tailwind, postcss-loader, and autoprefixer; [exactly as the tailwind docs say](https://tailwindcss.com/docs/installation#install-tailwind-via-npm) (accounting for the fact that we install postcss-loader, not postcss):

```
tailwindcss@latest postcss-loader@latest autoprefixer@latest
```

And everything just worked. So I'll keep digging, but it seems to me so far that upgrading css-loader might be all that's required.

**Comment by jtoar** [comment]
More to the point that we don't have it installed: https://github.com/redwoodjs/redwood/pull/696.

**Comment by thedavidprice** [comment]
@jtoar Ah, great work figuring this out. I assumed it must be imported having seen it used in Webpack, but it makes sense we need to have the Webpack config in the case someone adds it to their project.

I'm not sure when @dac09 and @peterp will be able to help me resolve the issues [discussed in this comment](https://github.com/redwoodjs/redwood/pull/2402#issuecomment-844620019). If you want to move ahead with a separate PR that updates `css-loader` along with the Tailwind setup command, feel free to go for it.

**Comment by thedavidprice** [comment]
FYI: Storybook v6.3 will include stable support for Webpack v5
https://github.com/storybookjs/storybook/issues/14397

**Comment by thedavidprice** [comment]
@peterp I attempted further diagnoses. I believe it's a Webpack config error. I attempted disabling loaders and plugins and ... to no avail — I never could resolve the error or produce a different one. My 🧠 is 🍳

**Comment by thedavidprice** [comment]
this has moved to #2884

Keeping this open for reference in timebeing
